From: Steve Best <sbest@redhat.com>
Date: Thu, 18 Feb 2010 01:16:06 -0500
Subject: [net] ixgbe: prevent speculatively processing descriptors
Message-id: <20100218010839.12835.49866.sendpatchset@squad5-lp1.lab.bos.redhat.com>
Patchwork-id: 23325
O-Subject: [PATCH RHEL5.5 BZ566309] ixgbe: prevent speculative processing of
	descriptors
Bugzilla: 566309
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: David Howells <dhowells@redhat.com>

RHBZ#:
======
https://bugzilla.redhat.com/show_bug.cgi?id=566309

Description:
============
The PowerPC architecture does not require loads to independent bytes to be
ordered without adding an explicit barrier.

In ixgbe_clean_rx_irq we load the status bit then load the packet data.
With packet split disabled if these loads go out of order we get a
stale packet, but we will notice the bad sequence numbers and drop it.

The problem occurs with packet split enabled where the TCP/IP header and data
are in different descriptors. If the reads go out of order we may have data
that doesn't match the TCP/IP header. Since we use hardware checksumming this
bad data is never verified and it makes it all the way to the application.

This bug was found during stress testing and adding this barrier has been shown
to fix it.

Signed-off-by: Milton Miller <miltonm@bga.com>
Signed-off-by: Anton Blanchard <anton@samba.org>

RHEL Version Found:
===================
RHEL 5.5

kABI Status:
============
No symbols were harmed.

Brew:
=====
http://brewweb.devel.redhat.com/brew/taskinfo?taskID=2264872

Upstream Status:
================
posted here
http://marc.info/?l=linux-netdev&m=126593061601526&w=
Intel plans to ack upstream once tests are completed successfully.

Test Status:
============
Tested by IBM test team
Intel QA has this patch under test right now. Test results should be
available soon.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ixgbe/ixgbe_main.c b/drivers/net/ixgbe/ixgbe_main.c
index e9eaa04..73bcdd4 100644
--- a/drivers/net/ixgbe/ixgbe_main.c
+++ b/drivers/net/ixgbe/ixgbe_main.c
@@ -807,6 +807,7 @@ static bool ixgbe_clean_rx_irq(struct ixgbe_q_vector *q_vector,
 			break;
 		(*work_done)++;
 
+		rmb(); /* read descriptor and rx_buffer_info after status DD */
 		if (rx_ring->flags & IXGBE_RING_RX_PS_ENABLED) {
 			hdr_info = le16_to_cpu(ixgbe_get_hdr_info(rx_desc));
 			len = (hdr_info & IXGBE_RXDADV_HDRBUFLEN_MASK) >>
